segmentation/utils/spatial_analysis.py

```diff
@@ -0,0 +1,245 @@
+                           dist_mat, marker_num, dist_lim, cell_label_idx):
+    """Finds positive cell labels and creates matrix with counts for cells positive for corresponding markers.
+
+    Args:
+        patient_data: cell expression data for the specific point
+        patient_data_markers: cell expression data of markers
+            for the specific point
+        thresh_vec: matrix of thresholds column for markers
+        dist_mat: cells x cells matrix with the euclidian
+            distance between centers of corresponding cells
+        marker_num: number of markers in expresion data
+        dist_lim: threshold for spatial enrichment distance proximity
+        cell_label_idx: column with cell labels
+
+    Returns:
+        close_num: marker x marker matrix with counts for cells
+            positive for corresponding markers
+        marker1_num: list of number of cell labels for marker 1
+        marker2_num: list of number of cell labels for marker 2"""
+
+    # Create close_num, marker1_num, and marker2_num
+    close_num = np.zeros((marker_num, marker_num), dtype='int')
+    marker1_num = []
+    marker2_num = []
+
+    for j in range(0, marker_num):
+        # Identify cell labels that are positive for respective markers
+        marker1_thresh = thresh_vec.iloc[j]
+        marker1posinds = patient_data_markers[patient_data_markers.columns[j]] > marker1_thresh
+        marker1poslabels = patient_data.loc[marker1posinds, patient_data.columns[cell_label_idx]]
+        marker1_num.append(len(marker1poslabels))
+        for k in range(0, marker_num):
+            # Identify cell labels that are positive for above marker and all other markers
+            marker2_thresh = thresh_vec.iloc[k]
+            marker2posinds = patient_data_markers[patient_data_markers.columns[k]] > marker2_thresh
+            marker2poslabels = patient_data.loc[marker2posinds, patient_data.columns[cell_label_idx]]
+            marker2_num.append(len(marker2poslabels))
+            # Subset the distance matrix to only include positive cell labels
+            trunc_dist_mat = dist_mat[np.ix_(
+                np.asarray(marker1poslabels - 1), np.asarray(
+                    marker2poslabels - 1))]
+            # Binarize the truncated distance matrix to only include cells within distance limit
+            trunc_dist_mat_bin = np.zeros(trunc_dist_mat.shape, dtype='int')
+            trunc_dist_mat_bin[trunc_dist_mat < dist_lim] = 1
+            # Record the number of interactions and store in close_num in index corresponding to both markers
+            close_num[j, k] = np.sum(np.sum(trunc_dist_mat_bin))
+    return close_num, marker1_num, marker2_num
+
+
+def compute_close_cell_num_random(marker1_num, marker2_num, patient_data,
+                                  dist_mat, marker_num, dist_lim, cell_label_idx, bootstrap_num):
+    """Uses bootstrapping to permute cell labels randomly.
```

Comment:
It permutes them randomly, and also records the number of close cells in that random setup. I think that's a key point here